### PR TITLE
Handle default expression value properly in Value content analysis

### LINKED ISSUE #6164: CA1508 False positive for "'x != null' is always 'true'"
### Steps To Reproduce

```csharp
#define REPRO

public static void Bar(ConsoleColor c)
{
    var color = default(string?);

    switch (c)
    {
        case ConsoleColor.Red:
        case ConsoleColor.DarkRed:
            color = "red";
            break;
        case ConsoleColor.Blue:
        case ConsoleColor.DarkBlue:
            color = "blue";
            break;
        default:
#if !REPRO
            color = null;
#endif
            break;
    }

    if (color != null)
#if REPRO
//      ~~~~~
//      CA1508: 'color != null' is always 'true'
#endif
    {
        // stuff
    }
}
```

### Expected behavior

No warning.

### Actual behavior

Issues warning.

## Additional context

Note that Roslyn also reports nullable analysis info in the tooltip stating "'color' may be null here.", so the analyzer and compiler have different opinions here.

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #6164

### CODE REVIEW

**Comment by mavasani** [comment]
@Youssef1313

File: src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs [hunk 1]
```diff
@@ -294,6 +294,11 @@ public static bool IsAttribute(this ITypeSymbol symbol)
         public static bool HasValueCopySemantics(this ITypeSymbol typeSymbol)
             => typeSymbol.IsValueType || typeSymbol.SpecialType == SpecialType.System_String;
 
+        public static bool HasReferenceTypeSemantics([NotNullWhen(returnValue: true)] this ITypeSymbol? typeSymbol)
+            => typeSymbol.IsReferenceTypeOrNullableValueType() ||
```
Comment (Youssef1313) [inline src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs:298]: The method name is quite unclear to me. For example, what does it mean that a nullable value type has reference type semantics?

**Review by Youssef1313 — APPROVED** [review]

File: src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs (hunk 1 above)
Comment (mavasani) [reply src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs:298]: @Youssef1313 - do you have a recommended name suggestion, I couldn't think of a good name to indicate `is default value for this type 'null'`?

Comment (Youssef1313) [reply src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs:298]: Would `CanHoldNullValue` be good?

Comment (Youssef1313) [reply src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs:298]: Maybe we should even revise other usages of `IsReferenceTypeOrNullableValueType` and maybe move the extra cases in it?

Comment (mavasani) [reply src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs:299]: > Maybe we should even revise other usages of `IsReferenceTypeOrNullableValueType` and maybe move the extra cases in it?

Let me file a separate issue to audit the callsites, and make fixes with appropriate unit tests being added.

**Pull request merged**